### PR TITLE
Feature/dependency tree

### LINKED ISSUE #27: 🐛 Minstall Does not Work With NPM5
**Problem**

The current version of minstall is not working correctly with NPM5.

**Analysis**

This is due to a changed behaviour in the npm install logic. For further reference see: #24.

In short:

We do a second `npm install` with minstall; 
this deletes the packages, that were installed during the first run.

**Solution**

1. Use a install mechanism that does not use `npm install` a second time.

   Just use a complete installation; link the rest.

### PR DESCRIPTION
**Pull request by heikomat**
See [the 3.0.0 changelog](https://github.com/heikomat/minstall/blob/6c786302204156bf59631cf94960c4733360399a/CHANGELOG.md#300) for an overview of what has changed.

This fixes #27

### CODE REVIEW

File: obsolete_functions.js [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+// minstall.js
+// this method isn't needed because npm4 doesn't create dead symlinks
+// and incompatible versions created through symlinks are cleaned up
+// in removeContradictingInstalledDependencies.
+// The dead symlinks created with npm5 are '--clean'ed pre-install anyway.
+function removeDeadSymlinksInInstalledDependencies() {
+  let folderPaths = [];
+  let flattenedFullFilePathArray = [];
+  let flattenedScopedFolderPaths = [];
+  let folderToCheck = [];
```
Comment (cmg-dev) [inline obsolete_functions.js:12]: The namings are inconsitent.

I would suggest just to call everything `*Array`.

Comment (cmg-dev) [reply obsolete_functions.js:12]: just pick a consistent pattern.

File: obsolete_functions.js [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+      });
+
+      const scopedFolderPaths = filesArray.map((files, index) => {
+        return files.filter((filename) => {
+          return filename.startsWith('@');
+        }).map((filename) => {
+          return path.join(folderPaths[index], filename);
+        });
+      });
+
```
Comment (cmg-dev) [inline obsolete_functions.js:43]: superfluous newline.

File: obsolete_functions.js [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+      flattenedScopedFolderPaths = [].concat.apply([], scopedFolderPaths);
+
+      return Promise.all(flattenedScopedFolderPaths.map((scopedFolderPath) => {
+        return systools.readdir(scopedFolderPath);
+      }));
+    })
+    .then((scopedDependencyFolder) => {
+      const fullScopedFolderPaths = scopedDependencyFolder.map((files, index) => {
+        return files.filter((filename) => {
+          return !filename.startsWith('.');
```
Comment (cmg-dev) [inline obsolete_functions.js:56]: yeah, I don't like this.

Because the function is unnamed, I do not have a chance to see if this is correct or, even worse, what the intention is. The intention is hidden in the return statement and is not semantically named.

File: obsolete_functions.js [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+  return new Promise((resolve, reject) => {
+    fs.move(from, to, options, (error) => {
+      if (error) {
+        return reject(error);
+      }
+
+      return resolve();
+    });
+  });
+},
```
Comment (cmg-dev) [inline obsolete_functions.js:187]: I suggest moving this to a GIST.

File: lib/moduletools.js [hunk 5]
```diff
@@ -30,14 +35,65 @@ const moduletools = {
...
+    const result = {
+      modules: [],
+      installedDependencies: [],
+    };
+
+    // get the local modules and the installed modules within the current path
+    return Promise.all([
+      ModuleInfo.loadFromFolder(location, ''),    // get infos about the current module
+      this.getModules(location, 'node_modules'),  // get infos about installed modules in the current module
+      this.getModules(location, folderName),      // get infos about local submodules of the current module
```
Comment (cmg-dev) [inline lib/moduletools.js:56]: You can spare the comments, if you extract this into functions and give them a proper name.

File: lib/minstall.js [hunk 6]
```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
...
-        return systools.link(modulePath, targetPath);
-      }));
+  const logLevels = {
+    critical: {level: 0, color: 'red'},
+    error: {level: 1, color: 'magenta'},
+    warn: {level: 2, color: 'yellow'},
+    info: {level: 3, color: 'green'},
+    verbose: {level: 4, color: 'gray'},
+    debug: {level: 5, color: 'blue'},
+    silly: {level: 6, color: 'cyan'},
```
Comment (cmg-dev) [inline lib/minstall.js:139]: yeah I found something I do not like ;)

Comment (cmg-dev) [reply lib/minstall.js:139]: do you need everyone of this level?

Please drop those you do not need.

Comment (cmg-dev) [reply lib/minstall.js:139]: https://github.com/heikomat/minstall/pull/28/files#diff-aab1c68f02682c785afd35196f518d24R124 

I guess only three are needed.

File: lib/minstall.js [hunk 7]
```diff
@@ -34,48 +39,28 @@ function logIfInRoot(message) {
...
-  });
+  return ModuleInfo.loadFromFolder(cwd, '');
 }
 
 function checkStartConditions() {
   logger.debug('checking start conditions');
   return Promise.all([
     systools.verifyFolderName(cwd, 'node_modules'),
     getLocalPackageInfo(),
+    systools.runCommand('npm --version'),
```
Comment (cmg-dev) [inline lib/minstall.js:51]: I would ALWAYS prefer to use a map instead of a list.

File: lib/minstall.js [hunk 8]
```diff
@@ -34,48 +39,28 @@ function logIfInRoot(message) {
...
   ])
     .then((results) => {
       const folderName = results[0];
       localPackage = results[1];
+      npmVersion = results[2];
```
Comment (cmg-dev) [inline lib/minstall.js:56]: If you would have used a map, you would not get this ugly integer-access to array fields.

Additionally it would be much more robust, as it does not depend on the order of the list.

File: lib/minstall.js [hunk 9]
```diff
@@ -34,48 +39,28 @@ function logIfInRoot(message) {
...
+
+      // npm 5 workaround until npm-issue #16853 is fixed
+      if (semver.major(npmVersion) === 5) {
+        logger.info('npm 5 detected. forcing --cleanup');
+        cleanup = true;
+      }
```
Comment (cmg-dev) [inline lib/minstall.js:62]: you could extract the check to a function, then initialize the cleanup variable at start and make it `const`.

File: lib/minstall.js [hunk 10]
```diff
@@ -34,48 +39,28 @@ function logIfInRoot(message) {
...
+
       const pathParts = cwd.split(path.sep);
       let parentFolder = pathParts[pathParts.length - 2];
```
Comment (cmg-dev) [inline lib/minstall.js:65]: `pathParts.length - 2`

Thats basically a magic number.

Consider putting this into a function? giving this a kind of business rule feel.

Why is the specific place in the array used?

File: lib/minstall.js [hunk 11]
```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
...
+        requestedDependencies[dependency] = {};
+      }
 
-function runPostinstall(moduleInfo) {
-  if (!moduleInfo.postinstall) {
-    logger.debug('won\'t run postinstall because there is no postinstall script');
-    return Promise.resolve();
-  }
+      let versionFound = false;
+      for (const version in requestedDependencies[dependency]) {
```
Comment (cmg-dev) [inline lib/minstall.js:164]: Where does `dependency` come from here?

File: lib/minstall.js [hunk 12]
```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
...
+        let intersection = null;
+        try {
+          intersection = intersect(requestedVersion, version);
+        } catch (error) {
+          // the versions didn't intersect. That's ok!
```
Comment (cmg-dev) [inline lib/minstall.js:169]: it will raise an error if the versions do not intersect?

What is a version Intersection?

Comment (cmg-dev) [reply lib/minstall.js:169]: probably call `error`: `noIntersectingVersionsFoundError`

File: lib/minstall.js [hunk 13]
```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
...
+        if (intersection) {
+          if (intersection !== version) {
+            requestedDependencies[dependency][intersection] = requestedDependencies[dependency][version];
+            delete requestedDependencies[dependency][version];
+          }
 
-      let command = moduleInfo.postinstall;
-      if (command.indexOf('minstall') === 0) {
-        command += ' --isChildProcess';
+          requestedDependencies[dependency][intersection].push(module.fullModulePath);
```
Comment (cmg-dev) [inline lib/minstall.js]: I would like to have this structure documented!

What is `requestedDependencies`. What is the "model" you expect it to have?

File: lib/minstall.js [hunk 14]
```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
...
+  for (const requestedDependencyName in result) {
+    for (const requestedDependencyVersionRange in result[requestedDependencyName]) {
 
-  if (conflictedDependencies.length === 0) {
-    logger.debug('there are no conflicting packages, thus nothing needs to be cached');
-    return Promise.resolve([]);
-  }
+      // check if already installed modules satisfy the dependency
+      for (const installedDependency of alreadyInstalledDependencies) {
+        if (requestedDependencyName !== installedDependency.name) {
```
Comment (cmg-dev) [inline lib/minstall.js:204]: We should add a refactor `minstall` issue to purge those if blocks.

File: lib/minstall.js [hunk 15]
```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
...
-    cachedModuleFolders = cachedModuleFolders.filter((module) => {
-      return localModuleNames.indexOf(module.name) < 0;
-    });
-  }
+        for (const locationOfRequestingModule of result[requestedDependencyName][requestedDependencyVersionRange]) {
+          // It doesn't matter where the dependency with the correct version is! all shadowed dependencies
+          // get fixed with symlinks after the installation!
+          const shortenedRequesterLocation = locationOfRequestingModule.substr(cwd.length);
+          const shortenedInstalledLocation = installedDependency.location.substr(cwd.length);
+          logger.debug(`dependency ${requestedDependencyName}@${requestedDependencyVersionRange} requested by '${shortenedRequesterLocation}' will be satisfied by installed version ${installedDependency.version} in '${path.join(shortenedInstalledLocation)}'`);
```
Comment (cmg-dev) [inline lib/minstall.js:217]: I would plainly just use a log to info and error.

We could consider dropping debug-loglevel entirely.

File: lib/minstall.js [hunk 16]
```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
...
-  });
+        delete result[requestedDependencyName][requestedDependencyVersionRange];
+        break;
+      }
 
-  if (cachedModuleFolders.length === 0) {
-    logger.debug('there are no conflicting packages, thus nothing needs to be cached');
-    return Promise.resolve([]);
+      if (!result[requestedDependencyName][requestedDependencyVersionRange] ||
+        result[requestedDependencyName][requestedDependencyVersionRange].length === 0) {
```
Comment (cmg-dev) [inline lib/minstall.js]: Okay this is surely a thing to refactor.

Comment (cmg-dev) [reply lib/minstall.js]: Please add a comment, what the purpose/intention is here!

File: lib/minstall.js [hunk 17]
```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
...
+        delete result[requestedDependencyName][requestedDependencyVersionRange];
+        continue;
+      }
+
+      // check if local modules that will get linked satisfy the dependency
```
Comment (cmg-dev) [inline lib/minstall.js:230]: 👍

File: lib/minstall.js [hunk 18]
```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
...
+      if (!linkModules) {
+        continue;
+      }
+
+      for (const localModule of localModules) {
```
Comment (cmg-dev) [inline lib/minstall.js:235]: Comment missing, why is this here?

File: lib/minstall.js [hunk 19]
```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
...
+        if (requestedDependencyName !== localModule.name) {
+          continue;
+        }
+
+        if (!semver.satisfies(localModule.version, requestedDependencyVersionRange)) {
```
Comment (cmg-dev) [inline lib/minstall.js:240]: same here.

File: lib/minstall.js [hunk 20]
```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
...
+          continue;
+        }
+
+        for (const locationOfRequestingModule of result[requestedDependencyName][requestedDependencyVersionRange]) {
```
Comment (cmg-dev) [inline lib/minstall.js]: same here.

File: lib/minstall.js [hunk 21]
```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
...
-  if (cachedModuleFolders.length === 0 && conflictingModules.length === 0) {
-    logger.debug('there are no packages that need to be uncached');
-    return Promise.resolve();
+  for (const dependencyName in dependencies) {
+    for (const dependencyVersionRange in dependencies[dependencyName]) {
+      result.push({
+        name: dependencyName,
+        versionRange: dependencyVersionRange,
+        identifier: `${dependencyName}@"${dependencyVersionRange}"`,
+        requestedBy: dependencies[dependencyName][dependencyVersionRange],
```
Comment (cmg-dev) [inline lib/minstall.js:272]: 👍 for the map instead of a list.

File: lib/minstall.js [hunk 22]
```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
...
-  });
+  return result;
+}
 
-  logger.debug('uncaching previously cached packages');
-  const rootNodeModules = path.join(cwd, 'node_modules');
-  const cacheFolder = path.join(rootNodeModules, `${moduleFolderName}_cache`);
-  const moduleFolder = path.join(cwd, moduletools.modulesFolder, moduleFolderName);
-  const moduleNodeModules = path.join(moduleFolder, 'node_modules');
+function determineDependencyTargetFolder(requestedDependencyArray, alreadyInstalledDependencies) {
```
Comment (cmg-dev) [inline lib/minstall.js]: Inconsistend naming.

Assuming `requestedDependencyArray` and `alreadyInstalledDependencies` are both arrays.

File: lib/minstall.js [hunk 23]
```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
...
+  for (const requestedDependency of requestedDependencyArray) {
 
-      return systools.delete(path.join(cacheFolder));
+    // it doesn't matter where the dependency gets Installed, even if mutliple
+    // modules need it, because in that case it will be symlinked from wherever
+    // it is installed. Because of this we just work with the path of the first
+    // module that requests this dependency
+    const possiblePathElements = requestedDependency.requestedBy[0].substr(cwd.length).split(path.sep).filter((pathElement) => {
+      return pathElement.length > 0;
     });
```
Comment (cmg-dev) [inline lib/minstall.js:291]: put this on the to refactor list.

File: lib/minstall.js [hunk 24]
```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
...
+        if (conflictingDependency) {
+          logger.debug(`${requestedDependency.identifier} can't be installed to ${currentPath}. it'd conflict with the to be installed ${conflictingDependency.identifier}`);
+          installModuleHere = false;
+        }
+      }
+
+      if (!installModuleHere) {
+        currentPath = path.join(currentPath, possiblePathElement);
+        continue;
+      }
```
Comment (cmg-dev) [inline lib/minstall.js:343]: I would strongly expect just to see `continue` here.

Why is the currentPath set here?

Comment (cmg-dev) [reply lib/minstall.js:343]: and just in this case.

File: lib/minstall.js [hunk 25]
```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
...
 
-  const packetFolderNames = packetsToKeep.map((packet) => {
-    return packet.folderName;
+      logger.debug(`found a place to install ${requestedDependency.identifier}: ${currentPath}`);
+
+      // the dependency can be installed here :)
+      if (!optimalDependencyTargetFolder[currentPath]) {
```
Comment (cmg-dev) [inline lib/minstall.js]: also a good example, where a semantic variable name would clarify much of the intention here.

Should be refactored; I cannot see what is going on here.

File: lib/minstall.js [hunk 26]
```diff
@@ -452,8 +708,31 @@ function run() {
...
   }
+}
+
+function cleanupDependencies() {
+  return moduletools.getAllModulesAndInstalledDependenciesDeep()
+    .then((moduleInfos) => {
+      return Promise.all(moduleInfos.modules.map((moduleInfo) => {
+        // local modules should be linked using the folder-names they should have,
+        // no matter what folder-name they actually have, therefore don't use realFolderName here
+        return systools.delete(path.join(moduleInfo.fullModulePath, 'node_modules'));
```
Comment (cmg-dev) [inline lib/minstall.js:733]: refactor: Hard coded string.

**Review by cmg-dev — CHANGES_REQUESTED** [review]
Please have a look at the changes. I'm not entirely sure how much effort we should currently put into this.

File: lib/minstall.js (hunk 8 above)
Comment (heikomat) [reply lib/minstall.js:56]: using a map here is difficult. `Promise.all()` wants an Array as input, and has an array with the results in the same order as the input as result

File: lib/minstall.js (hunk 6 above)
Comment (heikomat) [reply lib/minstall.js:139]: I could live with reducing the loglevels to `error`, `warn`, `info` and `debug`, but the levels you found are just the ones that log to `stderr`. The levels that are not listed there log to `stdout`

File: lib/minstall.js (hunk 11 above)
Comment (heikomat) [reply lib/minstall.js]: from [line 157](https://github.com/heikomat/minstall/pull/28/files/971f05e54ff84ce2baf930c8ee3133e12fada605#diff-aab1c68f02682c785afd35196f518d24R157)

Comment (heikomat) [reply lib/minstall.js]: from [line 157](https://github.com/heikomat/minstall/pull/28/files/971f05e54ff84ce2baf930c8ee3133e12fada605#diff-aab1c68f02682c785afd35196f518d24R157)

File: lib/minstall.js (hunk 12 above)
Comment (heikomat) [reply lib/minstall.js:169]: I'm using the module [semver-intersect](https://www.npmjs.com/package/semver-intersect). i give it two semver-ranges as input, and it returns a semver-range that satisfies both input-ranges. if the input-ranges don't intersect, the module throws an Error.

The fact that the versions don't intersect is not an error from minstalls point of view though, therefore nothing happens here, and i put the comment that tells you that it's ok for versions to not intersect

File: lib/minstall.js (hunk 13 above)
Comment (heikomat) [reply lib/minstall.js]: You're right, it should be documented. fyi, the structure looks like this:

```
type locationOfModuleThatWantsThatDependency = string;

{
  [dependencyName: string]: {
    [requestedVersionRange: string]: Array<locationOfModuleThatWantsThatDependency>
  }
}

// example
requestedDependencies = {
   moment: {
     '^2.18.1': [
       '/Users/heiko/Desktop/Projekte/BPMN/process_engine_meta/timing'
     ]
   }
}
```

File: lib/minstall.js (hunk 15 above)
Comment (heikomat) [reply lib/minstall.js:217]: this can log hundrets of lines you most likely need when debugging, but don't care about when using minstall.

File: lib/minstall.js (hunk 24 above)
Comment (heikomat) [reply lib/minstall.js:343]: the three parts above this block check if there is any reason the module cannot be installed in the location that is currently checked. If any of these checks say "nope, can't install here", we need to check the next possible location (the next "currentPath"). because of this we set the currentPath to the next one that we can check and continue

File: lib/minstall.js (hunk 25 above)
Comment (heikomat) [reply lib/minstall.js]: You are correct. fyi:
`optimalDependencyTargetFolder[currentPath]` holds the info about all the dependencies that will get installed to the location `currentPath`.  If `currentPath` is not yet a member of the object `optimalDependencyTargetFolder`, it needs to be added before we can push to `optimalDependencyTargetFolder[currentPath]`. This is what this if-block does

File: obsolete_functions.js (hunk 3 above)
Comment (heikomat) [reply obsolete_functions.js:56]: you are correct. fyi: the intend is to filter folder-names that start with `.`, because these are definitely not modules, but some npm-internal folders like `.staging`. we must not touch these

**Comment by heikomat** [comment]
@cmg-dev i will no merge this, as it is already published and in use, so it should be on master. For your review to not get lost, I'll open an issue and reference the closed PR
